fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java

```diff
@@ -1322,8 +1322,54 @@ private void applyAccurals() {
 
     private void applyPeriodicAccruals(final Collection<LoanTransaction> accruals) {
         List<LoanRepaymentScheduleInstallment> installments = getRepaymentScheduleInstallments();
+        if (TemporaryConfigurationServiceContainer.getAccrualDateConfigForCharge().equalsIgnoreCase("submitted-date")) {
+            updateInstallmentsForChargeSubmittedDateProcessing(installments, accruals);
+        } else {
+            updateInstallmentsForDefaultProcessing(installments, accruals);
+        }
+        LoanRepaymentScheduleInstallment lastInstallment = getLastLoanRepaymentScheduleInstallment();
+        for (LoanTransaction loanTransaction : accruals) {
+            if (loanTransaction.getTransactionDate().isAfter(lastInstallment.getDueDate()) && !loanTransaction.isReversed()) {
+                loanTransaction.reverse();
+            }
+        }
+    }
+
+    private void updateInstallmentsForChargeSubmittedDateProcessing(List<LoanRepaymentScheduleInstallment> installments,
+            Collection<LoanTransaction> accruals) {
         for (LoanRepaymentScheduleInstallment installment : installments) {
+            Money interest = Money.zero(getCurrency());
+            Money fee = Money.zero(getCurrency());
+            Money penality = Money.zero(getCurrency());
+            for (LoanTransaction loanTransaction : accruals) {
+                LocalDate chargeDueDate = loanTransaction.getLoanChargesPaid().stream().findFirst().get().getLoanCharge().getDueDate();
```

Comment:
I think the logic is correct, however i would not have duplicate the rest of the logic instead:
```
boolean isBasedOnSubmittedOnDate = TemporaryConfigurationServiceContainer.getAccrualDateConfigForCharge().equalsIgnoreCase("submitted-date")
for (LoanRepaymentScheduleInstallment installment : installments) {
            Money interest = Money.zero(getCurrency());
            Money fee = Money.zero(getCurrency());
            Money penality = Money.zero(getCurrency());
            for (LoanTransaction loanTransaction : accruals) {
                LocalDate chargeDueDate = isBasedOnSubmittedOnDate ?loanTransaction.getLoanChargesPaid().stream().findFirst().get().getLoanCharge().getDueDate() : loanTransaction.getTransactionDate();
                boolean isInRange = installment.isFirstPeriod()
                        ? !chargeDueDate.isBefore(installment.getFromDate()) && !chargeDueDate.isAfter(installment.getDueDate())
                        : chargeDueDate.isAfter(installment.getFromDate()) && !chargeDueDate.isAfter(installment.getDueDate());
                if (isInRange) {
                    interest = interest.plus(loanTransaction.getInterestPortion(getCurrency()));
                    fee = fee.plus(loanTransaction.getFeeChargesPortion(getCurrency()));
                    penality = penality.plus(loanTransaction.getPenaltyChargesPortion(getCurrency()));
                    if (installment.getFeeChargesCharged(getCurrency()).isLessThan(fee)
                            || installment.getInterestCharged(getCurrency()).isLessThan(interest)
                            || installment.getPenaltyChargesCharged(getCurrency()).isLessThan(penality)
                            || (isInterestBearing() && getAccruedTill().isEqual(loanTransaction.getTransactionDate())
                                    && !installment.getDueDate().isEqual(getAccruedTill()))) {
                        interest = interest.minus(loanTransaction.getInterestPortion(getCurrency()));
                        fee = fee.minus(loanTransaction.getFeeChargesPortion(getCurrency()));
                        penality = penality.minus(loanTransaction.getPenaltyChargesPortion(getCurrency()));
                        loanTransaction.reverse();
                    }

                }
            }
            installment.updateAccrualPortion(interest, fee, penality);
        }
    }
```

@ruchiD What do you think?